src/Plugin/Field/FieldType/EntityReferenceOverrideItem.php

```diff
@@ -0,0 +1,167 @@
+<?php
+
+namespace Drupal\entity_reference_override\Plugin\Field\FieldType;
+
+use Drupal\Component\Utility\NestedArray;
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Field\FieldStorageDefinitionInterface;
+use Drupal\Core\Field\Plugin\Field\FieldType\EntityReferenceItem;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\TypedData\MapDataDefinition;
+
+/**
+ * Plugin implementation of the 'media' field type.
+ *
+ * @FieldType(
+ *   id = "entity_reference_override",
+ *   label = @Translation("Entity reference override"),
+ *   description = @Translation("An entity field containing an entity reference and additional data."),
+ *   category = @Translation("Reference"),
+ *   default_widget = "entity_reference_autocomplete",
+ *   default_formatter = "entity_reference_label",
+ *   list_class = "\Drupal\Core\Field\EntityReferenceFieldItemList"
+ * )
+ */
+class EntityReferenceOverrideItem extends EntityReferenceItem {
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function propertyDefinitions(FieldStorageDefinitionInterface $field_definition) {
+    $properties = parent::propertyDefinitions($field_definition);
+
+    $properties['overwritten_property_map'] = MapDataDefinition::create()
+      ->setLabel(t('Overwritten property map'));
+
+    return $properties;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function schema(FieldStorageDefinitionInterface $field_definition) {
+    $schema = parent::schema($field_definition);
+
+    $schema['columns']['overwritten_property_map'] = [
+      'description' => 'A map to overwrite media data per instance.',
+      'type' => 'blob',
+      'size' => 'big',
+      'serialize' => TRUE,
+    ];
+
+    return $schema;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __get($name) {
+    if ($name == 'entity' && !empty($this->values['overwritten_property_map'])) {
+      /** @var \Drupal\Core\Entity\ContentEntityInterface $entity */
+      $entity = parent::__get('entity');
+
+      $this->overwriteFields($entity, $this->values['overwritten_property_map']);
+      $translation = $entity->getTranslation($this->getLangcode());
+      $this->overwriteFields($translation, $this->values['overwritten_property_map']);
```

Comment:
This looks really odd - I'd have thought that $entity was already the expected translation. Also we applying the $this->values['overwritten_property_map']) to both - and these will be the translated values from the parent entity.

Hmmm.... translation is going to be funky because often entity references are not translatable. I.e. the reference is the same regardless of translation - but the correct translation will be loaded. The entity reference override field only needs to be translatable if the fields it's overriding are translatable - or maybe it should always be translatable. Hmmm....

Reply:
Would like to revisit all this in https://github.com/thunder/entity_reference_override/issues/4